### PR TITLE
Add metadata support to low-level Python framework

### PR DESCRIPTION
**Pull request by soltanmm**
@nathanielmanistaatgoogle Please review, but wait on merging because this code may or may not be validly putting metadata into a dictionary depending on how violation of key uniqueness may or may not be treated as UB (see #1160).

### CODE REVIEW

File: src/python/src/grpc/_adapter/_low_test.py [hunk 1]
```diff
@@ -115,6 +115,10 @@ def tearDown(self):
   def _perform_echo_test(self, test_data):
     method = 'test method'
     details = 'test details'
+    server_metadata_key = 'my_server_key'
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_low_test.py]: In the server->client direction there is both "leading" metadata ("headers") and "trailing" metadata ("trailers") - please exercise both kinds in this test (so, along with client->server metadata, three separate metadata values in total).

File: src/python/src/grpc/_adapter/_low_test.py [hunk 2]
```diff
@@ -115,6 +115,12 @@ def tearDown(self):
   def _perform_echo_test(self, test_data):
     method = 'test method'
     details = 'test details'
+    server_leading_metadata_key = 'my_server_leading_key'
+    server_leading_metadata_value = 'my_server_leading_value'
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_low_test.py]: Let's use binary values containing internal null bytes for all three metadata values.

Comment (murgatroid99) [inline src/python/src/grpc/_adapter/_low_test.py]: I'm pretty sure you aren't supposed to have binary metadata values unless the metadata key ends with `'-bin'`.

File: src/python/src/grpc/_adapter/_low_test.py [hunk 3]
```diff
@@ -115,6 +115,12 @@ def tearDown(self):
...
+    server_trailing_metadata_key = 'my_server_trailing_key'
+    server_trailing_metadata_value = 'my_server_trailing_value'
+    client_metadata_key = 'my_client_key'
+    client_metadata_value = 'my_client_value'
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_low_test.py]: Let's make this value... 2047 null bytes. :-)

Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_low_test.py]: 16376 unset bits. :-)

File: src/python/src/grpc/_adapter/_call.c [hunk 4]
```diff
@@ -236,6 +250,9 @@ static PyMethodDef methods[] = {
     {"complete", (PyCFunction)pygrpc_call_complete, METH_O,
      "Complete writes to this call."},
     {"accept", (PyCFunction)pygrpc_call_accept, METH_VARARGS, "Accept an RPC."},
+    {"add_metadata", (PyCFunction)pygrpc_call_add_metadata, METH_VARARGS,
+     "Add metadata to the call. May not be called after invoke or "
+     "premetadata."},
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_call.c]: The "or premetadata" smells stale here, given that calling add_metadata after premetadata is how terminal metadata gets added, right?

File: src/python/src/grpc/_adapter/_call.c [hunk 5]
```diff
@@ -160,8 +160,22 @@ static const PyObject *pygrpc_call_accept(Call *self, PyObject *args) {
...
+  }
+  grpc_metadata metadata;
+  metadata.key = key;
+  metadata.value = value;
+  metadata.value_length = value_length;
+  return pygrpc_translate_call_error(
+      grpc_call_add_metadata_old(self->c_call, &metadata, 0));
+}
+
 static const PyObject *pygrpc_call_premetadata(Call *self) {
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_call.c]: Feel perfectly free to rename the "premetadata" function(s). Actively working with this code now is there a name that occurs to you as more appropriate?

File: src/python/src/grpc/_adapter/_low_test.py (hunk 2 above)
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_low_test.py]: Can you find that formally specified anywhere?

File: src/python/src/grpc/_adapter/_call.c (hunk 5 above)
Comment (soltanmm) [inline src/python/src/grpc/_adapter/_call.c]: The only obvious naming schemes that come to mind end up also changing the names of all the other functions to have 'client' and 'server' variants, s.t. it becomes clearer that sending the initial metadata is explicitly a server-only thing by not having a client variant. Alternatively, more literally wrapping the C API s.t. the names on the Python type reflect those of the C type.

But, really, I'm ambivalent; this is all getting hidden from the public API anyway, and clarifying information would probably be better served in the doc string without making larger scale changes to nomenclature.

File: src/python/src/grpc/_adapter/_low_test.py (hunk 2 above)
Comment (soltanmm) [inline src/python/src/grpc/_adapter/_low_test.py]: This contradicts what I've heard offline; ostensibly metadata values have to both have a length value and be null-terminated. Yes, it's redundant, but that was my impression from offline conversations. This sounds like it needs a clarifying issue to me.

Comment (murgatroid99) [inline src/python/src/grpc/_adapter/_low_test.py]: I don't know if it's documented, but it's true: https://github.com/grpc/grpc/blob/2e165e34010e44f20c6ecea63b20a6996810313f/src/core/transport/chttp2/bin_encoder.c#L276

Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_low_test.py]: :-(

Alright, please add separate keys ending with "-bin" and values to this test case.

**Comment by nathanielmanistaatgoogle** [comment]
Looks good; will merge when test results are in. :-)

**Pull request merged**